File: services/diagnosis/diagnosis/events.py
"""CloudEvent emission helpers."""

from __future__ import annotations

import json
import uuid
from abc import ABC, abstractmethod
from datetime import datetime, timezone
from typing import Any


def utcnow() -> datetime:
    return datetime.now(timezone.utc)


def new_event(
    *,
    tenant_id: str | None,
    workspace_id: str | None,
    event_type: str,
    subject: str,
    data: dict[str, Any],
) -> dict[str, Any]:
    return {
        "specversion": "1.0",
        "id": str(uuid.uuid4()),
        "source": "forge://service/diagnosis",
        "type": event_type,
        "subject": subject,
        "time": utcnow().isoformat(),
        "datacontenttype": "application/json",
        "forgetenantid": tenant_id or "",
        "forgeworkspaceid": workspace_id or "",
        "data": data,
    }


class Sink(ABC):
    @abstractmethod
    def emit(self, event: dict[str, Any]) -> None: ...


class MemorySink(Sink):
    def __init__(self) -> None:
        self.events: list[dict[str, Any]] = []

    def emit(self, event: dict[str, Any]) -> None:
        self.events.append(event)


class LogSink(Sink):
    def emit(self, event: dict[str, Any]) -> None:
        print("event", json.dumps(event))

File: services/iac-drift/internal/drift/events.go
package drift

import (
	"encoding/json"
	"fmt"
	"time"
)

type CloudEvent struct {
	SpecVersion      string         `json:"specversion"`
	ID               string         `json:"id"`
	Source           string         `json:"source"`
	Type             string         `json:"type"`
	Subject          string         `json:"subject,omitempty"`
	Time             time.Time      `json:"time"`
	DataContentType  string         `json:"datacontenttype"`
	ForgeTenantID    string         `json:"forgetenantid,omitempty"`
	ForgeWorkspaceID string         `json:"forgeworkspaceid,omitempty"`
	Data             map[string]any `json:"data"`
}

type Sink interface {
	Emit(ev CloudEvent) error
}

type MemorySink struct {
	Events []CloudEvent
}

func (m *MemorySink) Emit(ev CloudEvent) error {
	m.Events = append(m.Events, ev)
	return nil
}

func (m *MemorySink) ByType(t string) []CloudEvent {
	var out []CloudEvent
	for _, e := range m.Events {
		if e.Type == t {
			out = append(out, e)
		}
	}
	return out
}

type LogSink struct{}

func (LogSink) Emit(ev CloudEvent) error {
	b, _ := json.Marshal(ev)
	fmt.Println("event", string(b))
	return nil
}

func newEvent(ws Workspace, t string, data map[string]any) CloudEvent {
	return CloudEvent{
		SpecVersion:      "1.0",
		ID:               newID(),
		Source:           "forge://service/iac-drift",
		Type:             t,
		Subject:          "iac-workspace/" + ws.ID,
		Time:             time.Now().UTC(),
		DataContentType:  "application/json",
		ForgeTenantID:    ws.TenantID,
		ForgeWorkspaceID: ws.WorkspaceID,
		Data:             data,
	}
}
